src/main/java/net/dv8tion/jda/internal/handle/GuildStickersUpdateHandler.java

```diff
@@ -0,0 +1,165 @@
+import net.dv8tion.jda.api.events.sticker.update.GuildStickerUpdateNameEvent;
+import net.dv8tion.jda.api.events.sticker.update.GuildStickerUpdateTagsEvent;
+import net.dv8tion.jda.api.utils.cache.CacheFlag;
+import net.dv8tion.jda.api.utils.data.DataArray;
+import net.dv8tion.jda.api.utils.data.DataObject;
+import net.dv8tion.jda.internal.JDAImpl;
+import net.dv8tion.jda.internal.entities.EntityBuilder;
+import net.dv8tion.jda.internal.entities.GuildImpl;
+import net.dv8tion.jda.internal.entities.sticker.GuildStickerImpl;
+import net.dv8tion.jda.internal.utils.Helpers;
+import net.dv8tion.jda.internal.utils.UnlockHook;
+import net.dv8tion.jda.internal.utils.cache.SnowflakeCacheViewImpl;
+import org.apache.commons.collections4.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class GuildStickersUpdateHandler extends SocketHandler
+{
+    public GuildStickersUpdateHandler(JDAImpl api)
+    {
+        super(api);
+    }
+
+    @Override
+    protected Long handleInternally(DataObject content)
+    {
+        if (!getJDA().isCacheFlagSet(CacheFlag.STICKER))
+            return null;
+        final long guildId = content.getLong("guild_id");
+        if (getJDA().getGuildSetupController().isLocked(guildId))
+            return guildId;
+
+        GuildImpl guild = (GuildImpl) getJDA().getGuildById(guildId);
+        if (guild == null)
+        {
+            getJDA().getEventCache().cache(EventCache.Type.GUILD, guildId, responseNumber, allContent, this::handle);
+            return null;
+        }
+
+        DataArray array = content.getArray("stickers");
+        List<GuildSticker> oldStickers, newStickers;
+        SnowflakeCacheViewImpl<GuildSticker> stickersView = guild.getStickersView();
+        EntityBuilder builder = api.getEntityBuilder();
+        try (UnlockHook hook = stickersView.writeLock())
+        {
+            TLongObjectMap<GuildSticker> stickersMap = stickersView.getMap();
+            oldStickers = new ArrayList<>(stickersMap.valueCollection()); //snapshot of sticker cache
+            newStickers = new ArrayList<>();
+            for (int i = 0; i < array.length(); i++)
+            {
+                DataObject current = array.getObject(i);
+                final long stickerId = current.getLong("id");
+                GuildStickerImpl sticker = (GuildStickerImpl) stickersMap.get(stickerId);
+                GuildStickerImpl oldSticker = null;
+
+                if (sticker == null)
+                {
+                    sticker = (GuildStickerImpl) builder.createRichSticker(current);
+                    newStickers.add(sticker);
+                }
+                else
+                {
+                    // sticker is in our cache which is why we don't want to remove it in cleanup later
+                    oldStickers.remove(sticker);
+                    oldSticker = sticker.copy();
```

Comment:
What triggers this event? Any add/update/remove? So do we get then entire list again on any change?
I'm assuming only one sticker can reasonable be added/updated/removed at a single time, so if we _do_ get the entire list on any change it means any change will end up cloning all existing stickers even if they didn't change.

Seems like an expensive pattern to me honestly. What value does this provide to us compared to the "if-statement + update value + dispatch event" pattern we have for other entities?

Reply:
This implementation is entirely copied from the handling of emoji updates. We copy the list to compare which ones were added/removed and firing respective events. I'm unsure why discord designed it this way.

Reply:
To be clear, making a shallow clone of the `valueCollection` makes sense for before/after comparison of add/removals.
I was talking about the `.copy()` to do attribute comparisons. Seems heavy handed compared to iteration and attribute checking, no?

Reply:
It allows us to handle remove/add/update in a simple method by just checking if oldSticker is null or not. Since sticker updates are quite rare and stickers themselves are very tiny, I think thats a reasonable thing to do.